### DISCUSSION TITLE
how to get the context variable inside of python script excute by a bash command line from a bash operator.

### DISCUSSION THREAD

**Discussion by whylovegithub** [apache/airflow#30826] [Q&A] [answered]
OK, let me explain the problem a little bit thoroughly.
I have a Bash Operator to use bash command to call a python script. So is there any method I can get the context variable inside that python script, like python operator, `provide_context = True`.
The cause of this is that I have a list of table names to excute the same sql command, just simply extract them all and put them inside od a s3 bucket and trigger a glue job. I want to do a try-except session so that in the future if some  table got deleted it will just pass that table and continuely finish the rest of the table names still in that list. But what I want is that I can send a notificaiton to my email, so that I know something goes wrong.
The team already has a email notification function require a airflow context variable. The function was designed to be called as Bash Operator's parameter `on_success_callback`  and  `on_failure_callback`.
Sorry about the plain text, I cant copy paste code but if you need any further information. I would love to provide.
Any instruction or general direction is appreciated.

**Comment by yelbom** [comment]
Curious about this as well

**Comment by potiuk** [comment, marked as answer]
No. There is no way, and there won't be -at least not directly.

Simply because you have to go through bash operator and passing a context would require either running in the same interpreter or serializing and deserializi g context so  that you can retrieve it in your Python Code.

This is not really possible in general case for BashOperator because BashOperator by definition runs a new process with Bash script and it does not have a way to serialize.or reserialize the context. 

Instead, you should use PythonOperator, PythonVirtualEnvOperator or ExternalPythonOperator and in case of VirtualEnv or ExernalPythonOperaror make sure that you have the same Python version and the same Airflow version installed in the target VirtualEnv. 

In case of PythonOperator - the context is passed 'as is' - as Python Object. In case of VirtualEnv/External operators it is serialized first in the airflow interpreter and then deserialized in the target interpreter when it starts your task. In those cases also your task callable has to be serializable as well so that the method of yours can be serialized in Airflow interpreter and deserialized in the target one.

This will allow you to pass context to a Python code run in a different interpreter - but NOT through a bash script.

If you need to execute some bash script, you can - if you like - execute it as subprocess in your task (so have your target interpreter to start a new process and execute Bash script there)  but you won't be able to pas Python context object there as well.

If you still want to use bash, you can pass 'some' context variable of your choice that can be converted to a string. You can - in your bash script use JINJA2 expression (same JINJA2 expression that you can use in templated fields) and make them part of your bash script - this is the way how you can pass stringified versions of chosen variables. You can see our examples of Bash script usage, they are stringifying some expressions from the context I believe.

It's all the matter of crossing boundaries between processes - Python -> Python you can serialize the context and deserialize (under the same Python, same Airflow condition). Python -> Bash -> Python you need to use Jinja and stringify your values to pass individually.

**Comment by potiuk** [comment]
BTW. My advice: Get rid of bash. There is nothing you can do in Bash that can't be done in Python. Rewrite your bash in Python. Last year we rewrote 10.000 lines Bash `breeze` development environment in Python. That was great decision. You won't regret rewriting your Bash in Python either.

**Reply by whylovegithub** [reply]
Thanks for the reply, much appreciated since I don't need to keep searching for this one.

**Reply by alexsuo** [reply]
"Getting rid of your old crap system and move on our exciting new platform"

Yet after so many years people are still on bash which explains something.
